core/domain/user_jobs_one_off.py

```diff
@@ -470,3 +471,142 @@ def map(user_settings_model):
+    missing.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.CollectionProgressModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        completed_activities_model = (
+            user_models.CompletedActivitiesModel.get_by_id(item.user_id))
+
+        if completed_activities_model is None:
+            completed_activities_model = (
+                user_models.CompletedActivitiesModel(id=item.user_id))
+            completed_activities_model.exploration_ids = (
+                item.completed_explorations)
+            completed_activities_model.put()
+            yield ('Regenerated Missing CompletedActivitiesModel', item.id)
+
+        if completed_activities_model.deleted:
+            completed_activities_model.deleted = False
+            completed_activities_model.put()
+            yield (
+                'Reverted delete status for CompletedActivitiesModel',
+                item.id)
+
+        col_model = collection_models.CollectionModel.get_by_id(
+            item.collection_id)
+
+        collection_node_ids = [
+            node['exploration_id'] for node in (
+                col_model.collection_contents['nodes'])]
+        exp_ids_to_remove = [
+            exp_id
+            for exp_id in item.completed_explorations if exp_id not in (
+                collection_node_ids)]
+
+        if exp_ids_to_remove:
+            item.completed_explorations = [
+                exp_id for exp_id in item.completed_explorations
+                if exp_id not in exp_ids_to_remove]
+            item.put()
+            yield (
+                'Invalid Exploration IDs cleaned from '
+                'CollectionProgressModel',
+                'Model id: %s, Collection id: %s, Removed exploration '
+                'ids: %s' % (
+                    item.id, item.collection_id, exp_ids_to_remove))
+
+    @staticmethod
+    def reduce(key, values):
+        values.sort()
+        yield (key, values)
+
+
+class CleanUpUserContributionsModelOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """Job that cleans up UserContributionsModel by removing deleted
+    explorations from user contribution.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.UserContributionsModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        fetched_created_exploration_model_instances = (
+            datastore_services.fetch_multiple_entities_by_ids_and_models(
+                [('ExplorationModel', item.created_exploration_ids)]))[0]
+
+        exp_ids_removed = []
+        for exp_id, exp_instance in list(python_utils.ZIP(
+                copy.deepcopy(item.created_exploration_ids),
+                fetched_created_exploration_model_instances)):
+            if exp_instance is None or exp_instance.deleted:
+                exp_ids_removed.append(exp_id)
```

Comment:
Could you point to the line of code that does this removal when an exploration is deleted in the future? (This is with the intent of making sure that no new errors occur.)

Reply:
Replied above.